### DISCUSSION TITLE
Comparison of Results from OpenPNM and DragonFly

### DISCUSSION THREAD

**Discussion by fsamaraie** [PMEAL/OpenPNM#2700] [Q&A]
Hello community!
I have been using DragonFly software to do some PNM. In the documentation of DragonFly it is stated that the their PNM is based on OpenPNM, so I started using OpenPNM to get better ideas about the calculations involved. However, I am unable to reproduce the results, possibly due to me making mistakes as I am new to python. As you can see in Image 1 (produced in jupyter) the PNM seems complex with many pores, while the figure in dragonfly(image 2) is more simple with fewer spheres. Additionally, in dragonfly I am getting a max equivalent diameter of 48, while I get a max pore diameter in jupyter = infinity. 
Can you please advise on where I am being wrong? I am attaching my code for reference. 

im=imageio.volread('blobs.tiff')
snow_output = ps.networks.snow2(im, voxel_size=1)
pn = op.io.network_from_porespy(snow_output.network)
ax = op.visualization.plot_coordinates(pn)
ax = op.visualization.plot_connections(pn, ax=ax)
from pprint import pprint
mods=op.models.collections.geometry.spheres_and_cylinders
pprint(mods)
pn.add_model_collection(mods)
pn.regenerate_models()
print(pn)
np.max(pn['pore.diameter'])

![1](https://user-images.githubusercontent.com/115000993/219873393-7ba45eb2-a4b4-4c8a-aa4a-4371ed8cd399.jpg)
![2](https://user-images.githubusercontent.com/115000993/219873394-f0ccb651-677b-4f02-a681-f26574304ce0.jpg)

**Comment by jgostick** [comment]
Spheres and cylinders can cause problems if the pores are large and close together so they overlap too much.  Try the cones and cylinders models.

**Comment by fsamaraie** [comment]
Thank you for your response sir @jgostick . I reworked the example with a simple model composed of spherical pores connected by a throat. This model was artificially generated in porespy, with sphere diameters = 20 unit and cylinder radius =2. Unfortunately, when I run the pnm modelling on it,  I get different values for the pores equivalent diameters( ~5 and 25~) as seen below. I have performed the same analysis on DragonFly, and it gave me the correct result (eqD ~ 20). As dragonfly is based on OpenPNM, I think I am doing a mistake but I cannot find it. Is there any conditioning that is required prior to running the modelling? 
I am attaching my code and an image of the artificial data that I generated and visualized in paraview. 
Thanks 

snow_output = ps.networks.snow2(im, voxel_size=1)
pn = op.io.network_from_porespy(snow_output.network)
a=op.models.geometry.pore_size.equivalent_diameter(pn,pore_volume='pore.volume', pore_shape='spheres')
np.min(a), np.max(a), np.average(a)
(5.49760513463253, 25.87530721774293, 15.686456176187729) 
![Screenshot 2023-02-23 132942](https://user-images.githubusercontent.com/115000993/220923926-670f7019-52b1-429b-9df4-e8928f1f448a.jpg)

**Reply by jgostick** [reply]
I'm sorry, but you're going to have to give me more context.  You're getting a max equivalent diameter of 25.8, but it should be 20?  Can you post a script that is a minimal representation of the problem?

**Comment by fsamaraie** [comment]
Hello sir, 
I am trying to find the pore equivalent diameter using open pnm. For this reason, I created an artificial image stack that contains only a single throat and two pores (stack is attached). As I created this stack myself, I know that the diameters of both spheres are 20 units. However, when I am running the script below, I am not getting these numbers. When I use the same stack on dragonfly, however, I get correct values. Dragonfly uses open pnm for pore network modelling. It seems that I do not know which function to call to calculate the equivalent diameter for a given network

My script is the as follows, and the pnm1.tif is attached. I was unable to upload a tif file so I compressed it .
I sincerely appreciate your help. 

im=imageio.volread('pnm1.tif')
snow_output = ps.networks.snow2(im, voxel_size=1)
pn = op.io.network_from_porespy(snow_output.network)
a=op.models.geometry.pore_size.equivalent_diameter(pn,pore_volume='pore.volume', pore_shape='spheres')
np.min(a), np.max(a), np.average(a)

I get the following answers
(5.49760513463253, 25.87530721774293, 15.686456176187729)

[pnm1.zip](https://github.com/PMEAL/OpenPNM/files/10815752/pnm1.zip)

**Comment by fsamaraie** [comment]
I think I discovered the mistake which was in me generating the artificial images. 
Thanks :D
